### PR TITLE
Support `[patch]` in crate_universe when using multiple `Cargo.toml`s

### PR DESCRIPTION
**Pull request by UebelAndre**
This change is intended to address the current issues in CI but requires some additional changes to be made.

- https://github.com/bazelbuild/rules_rust/issues/1858

Note that each commit is a unique change but they're all required to fix CI.

### CODE REVIEW

File: .bazelci/presubmit.yml [hunk 1]
```diff
@@ -440,17 +440,16 @@ tasks:
...
-  #   name: Crate Universe Examples
-  #   platform: macos
-  #   working_directory: examples/crate_universe
-  #   run_targets: *crate_universe_vendor_example_targets
-  #   build_targets:
-  #     - "//..."
-  #   test_targets:
-  #     - "//..."
-  #   build_flags: *aspects_flags
+  crate_universe_examples_macos:
```
Comment (UebelAndre) [inline .bazelci/presubmit.yml:443]: By replacing the use of `rules_foreign_cc` the issue mentioned in https://github.com/bazelbuild/rules_rust/issues/1814 is avoided

**Comment by UebelAndre** [comment]
> I'm a bit torn on this one - while I strongly prefer rustls to openssl, I have found it valuable in the past to point people to this example as an example of a complex C++-using crate... What's the motivation behind the replacement? Just working around the bazelci issue?

The use of `rules_foreign_cc` has lead to issues and is not something I want to weigh down these tests. If folks want an example for how to build a complicated C++-using crate I would suggest making one in the root of `./examples`. I think we have sufficient examples of this though unless you would like an example for specifically `openssl`

File: cargo/private/cargo_build_script_wrapper.bzl [hunk 2]
```diff
@@ -119,6 +119,9 @@ def cargo_build_script(
         build_script_kwargs["compatible_with"] = kwargs["compatible_with"]
         binary_kwargs.pop("compatible_with")
 
+    if "toolchains" in kwargs:
```
Comment (UebelAndre) [inline cargo/private/cargo_build_script_wrapper.bzl:122]: This addresses https://github.com/bazelbuild/rules_rust/issues/1859

File: crate_universe/src/splicing/splicer.rs [hunk 3]
```diff
@@ -296,6 +296,12 @@ impl<'a> SplicerKind<'a> {
         let installations =
             Self::inject_workspace_members(&mut manifest, manifests, workspace_dir)?;
 
+        // Collect all patches from the manifests provided
```
Comment (UebelAndre) [inline crate_universe/src/splicing/splicer.rs:299]: This addresses https://github.com/bazelbuild/rules_rust/issues/1858

File: examples/crate_universe/vendor_local_manifests/BUILD.bazel [hunk 4]
```diff
@@ -14,12 +14,22 @@ rust_library(
     aliases = aliases(),
     edition = "2018",
     proc_macro_deps = all_crate_deps(proc_macro = True),
+    # TODO: https://github.com/bazelbuild/rules_rust/issues/1860
```
Comment (UebelAndre) [inline examples/crate_universe/vendor_local_manifests/BUILD.bazel:17]: https://github.com/bazelbuild/rules_rust/issues/1860

File: crate_universe/src/splicing/splicer.rs [hunk 5]
```diff
@@ -471,6 +484,34 @@ impl<'a> SplicerKind<'a> {
...
+    fn inject_patches(manifest: &mut Manifest, patches: &cargo_toml::PatchSet) -> Result<()> {
+        for (registry, new_patches) in patches.iter() {
+            // If there is an existing patch entry it will need to be merged
+            if let Some(existing_patches) = manifest.patch.get_mut(registry) {
+                // Error out if there are duplicate patches
+                existing_patches.extend(
+                    new_patches
+                        .iter()
+                        .map(|(pkg, info)| {
+                            if existing_patches.contains_key(pkg) {
```
Comment (illicitonion) [inline crate_universe/src/splicing/splicer.rs:497]: It'd be nice to not error if they patches are identical

**Review by illicitonion — APPROVED** [review]
Makes sense - thanks!

File: crate_universe/src/splicing/splicer.rs (hunk 5 above)
Comment (UebelAndre) [reply crate_universe/src/splicing/splicer.rs:497]: Good idea, done!

**Pull request merged**